Improve performance in extract

eksperimental opened the pull request.

eksperimental commented on lib/extractable.ex:
```diff
@@ -53,44 +55,85 @@ defprotocol Extractable do
-    [elem | rest_list] = :maps.to_list(map)
-    rest = :maps.from_list(rest_list)
-    {:ok, {elem, rest}}
+    case Map.keys(map) do
```
OTP 21 introduced an iterator function where it allows us to get the first element in the map without traversing the whole term. If we are going to support exclusively Elixir v1.10+ this can be directly replace. In the meantime, we can detect the OTP version and implement this version of extract for maps.

eksperimental replied:
I looked into it, and it does traverse the whole map, and covert it into a nested tuple. So there is not benefit in using it.

Qqwy replied:
Thank you for looking into it. In that case, it definitely does not make sense to add conditional code.

eksperimental commented on lib/extractable.ex:
```diff
@@ -53,44 +55,85 @@ defprotocol Extractable do
+  @doc """
+  Extracts the first element to the Erlang term of ordering.
+  """
+  def extract(enumerable) do
```
this implementation can be simplified but just this

```elixir
{[element], rest} = Enum.split(%{a: 1, b: 2}, 1)
{element, rest}
```

We should do some benchmarking to see which one is more performant

Qqwy replied:
Obviously the actual implementation should be something like

```
case Enum.split(enumerable, 1) do
  {[], _} -> {:error, :empty}
  {[element], rest} -> {element, rest}
end
```

Not a bad idea, but we indeed need to benchmark this, because IIRC the reason I used `Enum.fetch` before was that I seem to remember that it short-circuits. But maybe that does not matter at all when used like this.

eksperimental commented on .tool-versions:
```diff
@@ -1 +1 @@
-elixir 1.5.0-rc.0
+elixir 1.7.0
```
this is the oldest currently supported Elixir version

eksperimental commented:
Benchmarks.

```
$ mix run benchmarks/extract.exs 
Operating System: Linux
Elixir 1.11.2
Erlang 23.1.4

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 2 s
memory time: 0 ns
parallel: 1
inputs: map_set_empty, map_set_100, map_set_10_000, map_set_1_000_000
Estimated total run time: 32 s

Benchmarking extract: Enum.fetch with input map_set_empty...
Benchmarking extract: Enum.fetch with input map_set_100...
Benchmarking extract: Enum.fetch with input map_set_10_000...
Benchmarking extract: Enum.fetch with input map_set_1_000_000...
Benchmarking extract: Enum.split with input map_set_empty...
Benchmarking extract: Enum.split with input map_set_100...
Benchmarking extract: Enum.split with input map_set_10_000...
Benchmarking extract: Enum.split with input map_set_1_000_000...

##### With input map_set_empty #####
Name                          ips        average  deviation         median         99th %
extract: Enum.split        8.80 M      113.67 ns ±16656.79%           0 ns         406 ns
extract: Enum.fetch        7.51 M      133.16 ns ±18436.76%           0 ns         471 ns

Comparison: 
extract: Enum.split        8.80 M
extract: Enum.fetch        7.51 M - 1.17x slower +19.49 ns

##### With input map_set_100 #####
Name                          ips        average  deviation         median         99th %
extract: Enum.split       89.40 K       11.19 μs  ±1239.40%       10.21 μs       18.54 μs
extract: Enum.fetch       77.42 K       12.92 μs  ±1130.02%       11.86 μs       21.55 μs

Comparison: 
extract: Enum.split       89.40 K
extract: Enum.fetch       77.42 K - 1.15x slower +1.73 μs

##### With input map_set_10_000 #####
Name                          ips        average  deviation         median         99th %
extract: Enum.split        777.53        1.29 ms    ±74.06%        1.22 ms        1.51 ms
extract: Enum.fetch        651.58        1.53 ms    ±66.50%        1.45 ms        2.33 ms

Comparison: 
extract: Enum.split        777.53
extract: Enum.fetch        651.58 - 1.19x slower +0.25 ms

##### With input map_set_1_000_000 #####
Name                          ips        average  deviation         median         99th %
extract: Enum.fetch          4.89      204.30 ms    ±20.83%      194.08 ms      319.61 ms
extract: Enum.split          4.43      225.86 ms    ±22.03%      219.16 ms      313.24 ms

Comparison: 
extract: Enum.fetch          4.89
extract: Enum.split          4.43 - 1.11x slower +21.56 ms
```

```elixir
defimpl Extractable, for: Any do
  @doc """
  Extracts the first element to the Erlang term of ordering.
  """
  def extract(enumerable) do
    case Enum.fetch(enumerable, 0) do
      :error ->
        {:error, :empty}

      {:ok, element} ->
        rest = Enum.drop(enumerable, 1)
        {:ok, {element, rest}}
    end
  end

  def extract_split(enumerable) do
    case Enum.split(enumerable, 1) do
      {[], _} -> {:error, :empty}
      {[element], rest} -> {element, rest}
    end
  end

  def extract_fetch(enumerable) do
    case Enum.fetch(enumerable, 0) do
      :error ->
        {:error, :empty}

      {:ok, element} ->
        rest = Enum.drop(enumerable, 1)
        {:ok, {element, rest}}
    end
  end
end
```

and the benchmark file.

```elixir
Benchee.run(
  [
    "extract: Enum.fetch": &Extractable.Any.extract_fetch(&1),
    "extract: Enum.split": &Extractable.Any.extract_split(&1)
  ],
  inputs: [
    map_set_empty: MapSet.new([]),
    map_set_100: MapSet.new(1..100),
    map_set_10_000: MapSet.new(1..10_000),
    map_set_1_000_000: MapSet.new(1..1_000_000)
  ],
  time: 2
  # memory_time: 2
)
```

Qqwy commented:
Interesting! It seems like `Enum.split` is faster for all but the largest MapSets, so we should probably use that one. :+1:

eksperimental commented:
> Interesting! It seems like `Enum.split` is faster for all but the largest MapSets, so we should probably use that one. +1

I run the tests for 10 second and also measured the memory usage. 
The speed results are the same, and the memory footprint seems to be slower for fetch.

```
Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 10 s
parallel: 1
inputs: map_set_empty, map_set_100, map_set_10_000, map_set_1_000_000
Estimated total run time: 2.93 min

Benchmarking extract: Enum.fetch with input map_set_empty...
Benchmarking extract: Enum.fetch with input map_set_100...
Benchmarking extract: Enum.fetch with input map_set_10_000...
Benchmarking extract: Enum.fetch with input map_set_1_000_000...
Benchmarking extract: Enum.split with input map_set_empty...
Benchmarking extract: Enum.split with input map_set_100...
Benchmarking extract: Enum.split with input map_set_10_000...
Benchmarking extract: Enum.split with input map_set_1_000_000...

##### With input map_set_empty #####
Name                          ips        average  deviation         median         99th %
extract: Enum.split       10.79 M       92.64 ns ±59300.21%           0 ns         306 ns
extract: Enum.fetch        6.07 M      164.80 ns ±35231.05%           0 ns         361 ns

Comparison: 
extract: Enum.split       10.79 M
extract: Enum.fetch        6.07 M - 1.78x slower +72.17 ns

Memory usage statistics:

Name                   Memory usage
extract: Enum.split           160 B
extract: Enum.fetch            96 B - 0.60x memory usage -64 B

**All measurements for memory usage were the same**

##### With input map_set_100 #####
Name                          ips        average  deviation         median         99th %
extract: Enum.split       90.87 K       11.01 μs   ±406.26%       10.06 μs       19.19 μs
extract: Enum.fetch       76.18 K       13.13 μs   ±508.32%       12.10 μs       22.86 μs

Comparison: 
extract: Enum.split       90.87 K
extract: Enum.fetch       76.18 K - 1.19x slower +2.12 μs

Memory usage statistics:

Name                   Memory usage
extract: Enum.split         8.76 KB
extract: Enum.fetch         9.65 KB - 1.10x memory usage +0.89 KB

**All measurements for memory usage were the same**

##### With input map_set_10_000 #####
Name                          ips        average  deviation         median         99th %
extract: Enum.split        772.43        1.29 ms    ±33.17%        1.23 ms        1.92 ms
extract: Enum.fetch        631.57        1.58 ms    ±30.95%        1.51 ms        2.38 ms

Comparison: 
extract: Enum.split        772.43
extract: Enum.fetch        631.57 - 1.22x slower +0.29 ms

Memory usage statistics:

Name                   Memory usage
extract: Enum.split       859.54 KB
extract: Enum.fetch       860.90 KB - 1.00x memory usage +1.36 KB

**All measurements for memory usage were the same**

##### With input map_set_1_000_000 #####
Name                          ips        average  deviation         median         99th %
extract: Enum.fetch          4.58      218.51 ms    ±14.99%      207.55 ms      349.85 ms
extract: Enum.split          4.42      226.00 ms    ±21.10%      216.04 ms      332.41 ms

Comparison: 
extract: Enum.fetch          4.58
extract: Enum.split          4.42 - 1.03x slower +7.48 ms

Memory usage statistics:

Name                   Memory usage
extract: Enum.fetch        91.55 MB
extract: Enum.split        82.15 MB - 0.90x memory usage -9.40205 MB

**All measurements for memory usage were the same**
```

eksperimental commented:
The one thing I don't like about the implementation for Any, is that we loose the original format of the enumerable, and we return a list. Is there anything we could do about?

Qqwy reviewed (commented):
> > The one thing I don't like about the implementation for Any, is that we loose the original format of the enumerable, and we return a list. Is there anything we could do to prevent this?
> 
> I guess with Enumerable lacking support for delete and put functions, this is not possible to do.

Yes; the whole reason `Extractable` exists is because `Enumerable`s can (in general) only be consumed all-at-once.

Qqwy commented:
> Actually, I'll go ahead and write a benchmark for this thinking

c.f. #14

eksperimental commented:
I am thinking we should provide the fall back to Any, because we need to rely on Enum.reduce for doing this operation

Qqwy commented:
> I am thinking we should provide the fall back to Any, because we need to rely on Enum.reduce for doing this operation

I don't quite understand your comment: There currently _is_ an implementation for Any which indeed only works for types that happen to be Enumerable, and which exists to be able to do 'something' with types that are Enumerable even though they will be converted to lists by it.

Were you considering removing that fallback before? Or something else?

eksperimental commented:
Sorry about the confusion, I meant to write "I am thinking we should **NOT** provide the fall back to Any"

eksperimental commented:
Closing in favor of already merged #17
